Update nf-handleapi-duplicatehandle.md

ptorr-msft opened the pull request:
Added security info on duplicating handles and included link to RPC attribute.

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

Added security info on duplicating handles and included link to RPC attribyute.

jonwis commented:
From reps discussion: Add a hedgeword to the Remarks section about what handles can be duplicated that this is a subset of known-safe-to-duplicate handles.

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

Made extra-clear that not all "handles" can be duped

ptorr-msft commented:
> From reps discussion: Add a hedgeword to the Remarks section about what handles can be duplicated that this is a subset of known-safe-to-duplicate handles.

The doc already mentions "don't try I/O ports or sockets" down below, but I added an extra-extra disclaimer as well.

jonwis commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -176,7 +183,7 @@ If the process that calls
 <b>DuplicateHandle</b> can be used to duplicate a handle between a 32-bit process and a 64-bit process. The resulting handle is appropriately sized to work in the target process. For more information, see <a href="https://docs.microsoft.com/windows/desktop/WinProg64/process-interoperability">Process Interoperability</a>.
 
-<b>DuplicateHandle</b> can duplicate handles to the following types of objects.
+Not all "handles" can be duplicated with <b>DuplicateHandle</b>. Only the following types of objects can have their handles duplicated:
```
Is this really the complete list? If so, the comment is good. Otherwise, consider replacing "Only the following" with "The list below is a subset of system handles that can be duplicated. Most kernel handles can be duplicated" or something similar.

keith-horton replied:
I don't think this is guaranteed to be a complete list - it's a safe list that is known to use NT file descriptors behind the HANDLE values returned to user-mode. I think Jon's quoted statement is more accurate.

oldnewthing replied:
I like the original text. "It can duplicate these types." Doesn't say the list is comprehensive. But also suggests that if your handle isn't on the list, you should check closer.

ptorr-msft replied:
I removed "only" so it's more like the original.

brialmsft commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -160,7 +164,10 @@ If the function fails, the return value is zero. To get extended error informati
+Handles should only be duplicated between two processes at the same privilege level (for example, between two processes running as the same user). Duplicating handles across privilege levels (for example, from a regular user to a service account) can cause security issues and should be avoided unless the implications are fully understood. Objects created with a default security descriptor will pr ...
+<a href="https://docs.microsoft.com/windows/desktop/ProcThread/process-security-and-access-rights">Process Security and Access Rights</a>.
 
+A more reliable, secure way to share object handles across processes is to use RPC and apply the **[system_handle]** attribute on the relevant method parameters. For more information, see <a href="/windows/win32/midl/system-handle">system_handle attribute</a>.
```
Might be better to say "COM or RPC" rather than just "RPC". Technically RPC covers both because COM's support for handle marshaling is built on RPC's. But it may be clearer to exernal devs if we're explicit that both stacks have system_handle support.

ptorr-msft replied:
Done

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

Added "COM" as well per Brian

brialmsft commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -354,6 +361,9 @@ Normally the target process closes a duplicated handle when that process is fini
 </ul>
 
+> [!Note] 
+> The **DUPLICATE_CLOSE_SOURCE** flag can close *any* handle inside the process specified by **hSourceProcessHandle**; it does not have to be a handle that was previously duplicated into the calling process. Unexpectedly closing a handle can cause reliability or security issues in the source process, so avoid using this flag unless you are sure the source process is expecting you to close the han ...
```
>  Typically, you should rely on the target process to close the handle.

I'm not following this. The rest of the note is talking about the source handle, and I don't think it's a common pattern for the target process to close the source handle.

ptorr-msft replied:
In a normal sequence the target process would get the duplicated handle and eventually call **CloseHandle** on it. It's weird to close a handle on behalf of another process (I assume there's a "good reason" why this is available as a feature....). I'll try rewording to: Typically the target process will close the handle via **CloseHandle**, so this flag isn't needed.

ptorr-msft replied:
(Note that the existing text above already says "Normally the target process closes a duplicated handle" so I am following that naming pattern explicitly).

brialmsft replied:
The target process typically closes the *duplicated* handle (aka the target handle) when it's done with it. My confusion comes from "you should rely on the target process to close the handle", where the only handle being talked about previously in this paragraph was the source handle. Overall I'd recommend not using "the handle" because it's ambiguous; it's clearer to always refer explicitly to the "source handle" vs the "target handle".

If you're talking about the target process's relationship to the target handle, well this note is addressing usage of the DUPLICATE_CLOSE_SOURCE handle; the lifetime of the target handle doesn't seem relevant here.

If you're talking about the target process's  relationship to the source handle, I would think it's atypical for the target process to close it. As I see it, the 2 basic patterns are:
* Push: The source process calls DuplicateHandle to push a handle from their process into the target process. The source process retains ownership of the source handle by default. If they want to close it immediately, they could use DUPLICATE_CLOSE_SOURCE (though they need to be careful that nobody else owns the handle and plans to use or close it).
* Pull: The target process calls DuplicateHandle to pull a handle from the source process into their process. The source process retains ownership of the source handle by default. It's certainly *possible* to have a usage pattern where the target process will close the source handle after the duplication, but I don't know if I'd call it "typical". And yes, they'd need to be careful about it.

Your "avoid using this flag unless you are sure the source process is expecting you to close the handle" pretty much covers the "be careful" aspect of both of those, so I'm not sure what the last sentence is trying to add to that.

ptorr-msft replied:
Updated with explicit "duplicate handle" (and updated original document text as well) since "duplicated" is ambiguous. I think the root of the problem is that the "source" process, when talking about the API, is the process identified by `hSourceProcessHandle` which _could_ be the original "source" of the duplicated handle, or it _could_ be the original "target" of the duplicate handle, or it _could_ be neither (just a random process). The entire API design is weird.

> If they want to close it immediately, they could use DUPLICATE_CLOSE_SOURCE 

Do we actually see anyone doing that rather than just calling `CloseHandle`? Seems super unintuitive to use **Duplicate** as a way to **Close**.

brialmsft replied:
Oh, I was dumb and read this Note without really reading the context of the preceding text. This is specifically about how to use DuplicateHandle as a remote CloseHandle, not general-purpose remarks on DUPLICATE_CLOSE_SOURCE. Apologies for the poor reading comprehension on my part!

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

per Brian's comment about CloseHandle

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -53,7 +53,7 @@ ms.custom: 19H1
 
 
-Duplicates an object handle.
+Duplicates an object handle. Duplicating object handles is recommended only between processes with the same privileges. Duplicating handles across privilege levels can introduce security vulnerabilities. See the **Remarks** section for more information.
```
I don't think we should be adding to the IntelliSense text. It's fine to add a reference to the Remarks from the parameter list, but I think the IntelliSense text should be brief.

ptorr-msft replied:
How do you get the intellisense to appear? I don't see anything in VS 2022.  Also this comment is much shorter than (eg) the comment for `lpTargetHandle`. I'll put int on a separate line just to be sure though.

ptorr-msft replied:
Simplified the header to just say there's important security info... which I think is nice for the Bing infocard as well.

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -160,7 +164,10 @@ If the function fails, the return value is zero. To get extended error informati
 
 ## -remarks
 
+Handles should only be duplicated between two processes at the same privilege level (for example, between two processes running as the same user). Duplicating handles across privilege levels (for example, from a regular user to a service account) can cause security issues and should be avoided unless the implications are fully understood. Objects created with a default security descriptor will pr ...
```
I think this is being alarmist. It is fairly standard practice for a high-privilege process to duplicate handles into or out of a low-privilege process.

ptorr-msft replied:
Yes, but you still have to be careful. I expanded it a bit - what do you think?

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -354,6 +361,9 @@ Normally the target process closes a duplicated handle when that process is fini
 </ul>
 
+> [!Note] 
+> The **DUPLICATE_CLOSE_SOURCE** flag can close *any* handle inside the process specified by **hSourceProcessHandle**; it does not have to be a handle that was previously duplicated into the calling process. Unexpectedly closing a handle can cause reliability or security issues in the source process, so avoid using this flag unless you are sure the source process is expecting you to close the han ...
```
I didn't see any an expectation the `DUPLICATE_CLOSE_SOURCE` could be used only to close handles that had previously been duplicated *into* the source process. "Close source" is basically a "move handle".

> Typically, you should rely on the target process to close the handle.

No, typically, you should rely on the *source* process to close the handle. The target process typically closes the duplicated handle, not the original handle.

brialmsft replied:
I had the exact same reaction to this (see my resolved comment chain). But this Note section is specifically referring to the pattern described just above where you effectively use DuplicateHandle as a remote CloseHandle. In this pattern there is no duplicated handle (lpTargetHandle is null). This last sentence is basically reiterating "remote close is weird, you probably shouldn't do it".

Since we both got tripped up on this, maybe it would be worth starting this Note with some lead-in text connecting it to the previous section?

ptorr-msft replied:
You might not have that expectation, but I'd bet you $1 that (at least) 9 out of 10 dentists would. Software Developers,  even :). Who would build an API called `DuplicateHandle` that lets you **close non-duplicated handles**? It's entirely bizarre.

As noted in the discussion with Brian, "source" vs "target" is very confusing in this context, but I am using 'target' as originally intended. I'll take another stab at being suuuuuper clear.

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -176,7 +183,7 @@ If the process that calls
+Not all "handles" can be duplicated with <b>DuplicateHandle</b>. Only the following types of objects can have their handles duplicated:
 
 <table>
 <tr>
```
Overall, I think we should address the broader picture, rather than how each individual flag fits into the big picture.

> Exercise caution when granting `PROCESS_DUP_HANDLE` access because the `DuplicateHandle` function can be used to manipulate the handles of the process in a way that leads to full control over that process. If two processes are running in different security contexts, it is generally better for the higher-privilege process to duplicate handles into or out of the lower-privilege process. For a more flexible and secure mechanism of duplicating handles between processes, see the **[system_handle]** RPC attribute.

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -366,21 +376,42 @@ DWORD CALLBACK ThreadProc(PVOID pvParam);
+    hMutex = CreateMutex(NULL, FALSE, NULL);
+    if (NULL == hMutex)
+    {
+      // Handle error here
```
Ambiguous comment. "Handle error here" could mean

* There is an error with the handle here. ("Handle" as a noun.)
* We should handle the error here. ("Handle" as a verb.)

The ambiguity is particularly strong since this function is about handles!

I think we can just delete the comment. The `OutputDebugStringW` explains what happened.

ptorr-msft replied:
Good point :)

oldnewthing commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -366,21 +376,42 @@ DWORD CALLBACK ThreadProc(PVOID pvParam);
 
+    if (FALSE == result)
+    {
+      // Handle error here
```
Maybe note `// Cleanup omitted for expository purposes`.

Because we are leaking the mutex (and the duplicated mutex) in all the error paths.

ptorr-msft replied:
Done. Should really be RAII but... want to keep it simple.

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

Re-write of various bits, trying to remove ambiguity about "source" and "target" in particular.

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

More typos etc.

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

smaillet-ms commented on sdk-api-src/content/handleapi/nf-handleapi-duplicatehandle.md:
```diff
@@ -342,16 +358,24 @@ The <i>dwDesiredAccess</i> parameter specifies the new handle's access rights. A
 
-Normally the target process closes a duplicated handle when that process is finished using the handle. To close a duplicated handle from the source process,  call <b>DuplicateHandle</b> with the following parameters: 
+In some cases, the duplicate handle can be granted more access rights than the original handle, but in other cases it cannot. For example, process handles can be duplicated with <b>PROCESS_ALL_ACCESS</b> rights regardless of the original handle's access rights, whereas a file handle created with the <b>GENERIC_READ</b> access right cannot be duplicated so that it has the <b>GENERIC_WRITE</b> acce ...
+
```
We really should spell out the details of this. It applies to the COM/RPC/ALPC marshaling as well, and sadly is about as clear as mud, even internally. In what scenarios will it work, in what scenarios will it silently be ignored, and in what scenarios will it fail. When dealing with a middle layer library you may not be the one opening the handle and can't make guarantees about the permissions it has, but want to ensure the target has at least X and NEVER Y. Clearly, there are rules on this sort of thing (and hopefully tests that validate them) we've just never done the hard work of collecting all the details and documenting it. Which is pretty bad from something security related.

ptorr-msft replied:
Agreed, but that seems like a Very Big Project and would be a separate topic (then we can refer to it from this page). Do we even know where to start? We had a hard time even figuring out which kinds of handles CAN be duplicated (never really resolved).

ptorr-msft committed:
Update nf-handleapi-duplicatehandle.md

Minor addition to the "close someone else's handle" bit

ptorr-msft commented:
Giving up for now,
